### PR TITLE
164796886 User Registration Mail Verification Feature

### PR DESCRIPTION
**Pull request by Simplemart17**
#### Description
When user signup, there is need to verify their account. This feature write function to send welcome mail and a verification link to confirm the email address.
#### Type of change
- New feature (non-breaking change which add functionality)
#### How Has This Been Tested?
- Integration testing
- Unit testing
#### Checklist:
- My code follows the style guidelines of this project
- My changes generate no new warning
- New and existing unit test pass locally with my changes
#### PT
[#164796886](https://www.pivotaltracker.com/story/show/164796886)

### CODE REVIEW

File: src/utils/email.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import sendGrid from '@sendgrid/mail';
+
+export default async (apiKey, payload) => {
+  sendGrid.setApiKey(apiKey);
+  const data = await sendGrid.send(payload);
+  return data;
+};
+
+
+// const msg = {
```
Comment (chikeozulumba) [inline src/utils/email.js]: @Simplemart17  Great work on this feature. I could not wrap my head around why these comments should be on file, would you consider resolving it by removing it or applying any other method as you see fit. Thanks

Comment (Simplemart17) [reply src/utils/email.js]: It has been resolved. Thanks

File: src/controllers/confirmation.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+   * @param {object}  next
+   * @returns {object|void} response object
+   */
+
+  async userVerification(req, res, next) {
+    try {
+      const { secretToken } = req.params;
+      const user = await User.findOne({
+        where: { secretToken }
+      },);
```
[the remark this answers is not in the document]
Comment (ssanusi) [reply src/controllers/confirmation.js]: @Simplemart17 please attend to houndCi

File: src/controllers/authentication/user.js [hunk 3]
```diff
@@ -9,52 +12,57 @@ import { responseFormat, errorResponseFormat } from '../../utils/index';
...
   try {
     const { body } = req;
-    const user = await User.create({ ...body });
+    const verificationToken = uuid();
+    const user = await User.create({ ...body, verificationToken });
     const { id, username, email, role, fullName, bio } = user;
 
+    let url = `http://${process.env.HOST}:${process.env.PORT}/api/v1/verify`;
+
+    if (process.env.NODE_ENV === 'production') { url = `https://${process.env.HOST}:${process.env.PORT}/api/v1/verify`; }
```
Comment (Temmyogunbo) [inline src/controllers/authentication/user.js]: Let us learn to separate code and put them where they should belong. This should not be here.

File: src/controllers/authentication/user.js [hunk 4]
```diff
@@ -9,52 +12,57 @@ import { responseFormat, errorResponseFormat } from '../../utils/index';
...
+    const verificationUrl = `${url}/${id}/${verificationToken}`;
+
+    const subject = 'Welcome to Authors\' Haven';
+
+    const message = `
+      <div>
+        <p>Hi ${fullName},</p>
+        <p>Welcome to Authors Haven, a place to be inspired! Your account was successfully created.</p>
+        <p>Please click this <a href=${verificationUrl}>link</a> to confirm your account.</p>
+      </div>`;
```
Comment (Temmyogunbo) [inline src/controllers/authentication/user.js]: This should not here.

File: src/controllers/authentication/user.js [hunk 5]
```diff
@@ -9,52 +12,57 @@ import { responseFormat, errorResponseFormat } from '../../utils/index';
...
-    return res.status(500).json(errorResponseFormat({
-      message: 'Something Went Wrong',
-    }));
+export const verifyUser = async (req, res) => {
+  try {
+    const { id, verificationToken } = req.params;
+    let user = await User.findOne({ where: { id, verificationToken } });
+    if (!user) { return res.status(403).json(errorResponseFormat({ status: 'fail', message: 'Invalid token supplied, kindly reauthenticate!', })); }
+    user = await user.update({ verified: true, verificationToken: null });
+    if (user) { return res.status(202).json(responseFormat({ status: 'success', data: 'Your account was successfully verified!', })); }
```
Comment (Temmyogunbo) [inline src/controllers/authentication/user.js]: Think of line 62 and 64, it is difficult to reason. Let us improve on the names

File: src/controllers/authentication/user.js (hunk 3 above)
Comment (Simplemart17) [reply src/controllers/authentication/user.js]: It has been resolved. Thanks

File: src/controllers/authentication/user.js (hunk 5 above)
Comment (Simplemart17) [reply src/controllers/authentication/user.js]: I have made it more readable. Thanks

**Review by Temmyogunbo — APPROVED** [review]

**Pull request merged**